library/types/test/url_test.rb

```diff
@@ -25,23 +25,123 @@
+          "scheme"   => "cd",
+          "host"     => "",
+          "path"     => "/",
+          "fragment" => "",
+          "user"     => "",
+          "pass"     => "",
+          "port"     => "",
+          "query"    => "device=/dev/sr0"
+        }
+      end
+
+      it "returns a hash containing 'scheme', 'path' and 'query'" do
+        expect(subject.Parse(url)).to eq(tokens)
+      end
+    end
+
+    context "given a CD/DVD with a device and a path" do
+      let(:url) { "cd:/some/file?device=/dev/sr0" }
+      let(:tokens) do
+        {
+          "scheme"   => "cd",
+          "host"     => "",
+          "path"     => "/some/file",
+          "fragment" => "",
+          "user"     => "",
+          "pass"     => "",
+          "port"     => "",
+          "query"    => "device=/dev/sr0"
+        }
+      end
+
+      it "returns a hash containing 'scheme', 'path' and 'query'" do
+        expect(subject.Parse(url)).to eq(tokens)
+      end
+    end
   end
 
   describe ".Build" do
     it "returns the URL for the given tokens" do
       expect(subject.Build(tokens)).to eq(url)
     end
 
-    context "given a cd/dvd URL" do
-      let(:tokens) do
-        {
-          "scheme" => "cd",
-          "query"  => "device=/dev/sr0"
-        }
+    context "given CD/DVD tokens including a device" do
+      context "with a device" do
+        let(:tokens) do
+          {
+            "scheme" => "cd",
+            "query"  => "device=/dev/sr0",
+            "path"   => "/"
+          }
+        end
+
+        it "returns a URL of the form 'cd:///?<device>'" do
+          expect(subject.Build(tokens)).to eq("cd:///?device=/dev/sr0")
+        end
+      end
+
+      context "with a directory" do
+        let(:tokens) do
+          {
+            "scheme" => "cd",
+            "path"   => "/dir"
+          }
+        end
+
+        it "returns a URL of the form 'cd:///<dir>'" do
+          expect(subject.Build(tokens)).to eq("cd:///dir")
+        end
       end
+    end
+  end
+
+  describe "URLs rebuilding" do
+    # This intention of these tests is to check if URLs are rebuilt correctly.
+
+    URLS = {
+      "dvd:/dir"                        => "dvd:///dir",
+      "dvd://dir"                       => "dvd:///dir",
+      "dvd:///dir"                      => "dvd:///dir",
+      "cd:/?device=/dev/sr0"            => "cd:///?device=/dev/sr0",
+      "cd:/some/file?device=/dev/sr0"   => "cd:///some/file?device=/dev/sr0",
+      "cd:///some/file?device=/dev/sr0" => "cd:///some/file?device=/dev/sr0",
+      "http://u:p@suse.de/a#b"          => "http://u:p@suse.de/a#b",
+      "ftp://u:p@suse.de/a#b"           => "ftp://u:p@suse.de/a#b",
+      "slp:/"                           => "slp://"
+    }.freeze
 
-      it "returns a URL containing which a single slash to separate the schema from the rest" do
-        expect(subject.Build(tokens)).to eq("cd:/?device=/dev/sr0")
+    URLS.each do |url, rebuilt|
+      context "given #{url}" do
```

Comment:
NP: IMHO the context is misused here, the context should define the environment where the test is run (something like "in installation", "on S390",...).

The test input should be part of the `it`, e.g. `it "returns '#{rebuilt}' for '#{url}'" do ...`